File: backend/app/routers/audit_logs.py
import math
from datetime import datetime
from typing import Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.dependencies import get_db, require_admin
from app.models.audit_log import AuditLog
from app.models.user import User
from app.schemas.audit_log import AuditLogListResponse, AuditLogResponse

router = APIRouter(prefix="/audit-logs", tags=["audit-logs"])


@router.get("", response_model=AuditLogListResponse)
def list_audit_logs(
    page: int = 1,
    page_size: int = 50,
    user_id: Optional[int] = Query(None),
    action: Optional[str] = Query(None),
    entity_type: Optional[str] = Query(None),
    date_from: Optional[datetime] = Query(None),
    date_to: Optional[datetime] = Query(None),
    db: Session = Depends(get_db),
    _admin: User = Depends(require_admin),
):
    """List audit logs with pagination and filters (admin only)."""
    query = db.query(AuditLog)

    if user_id is not None:
        query = query.filter(AuditLog.user_id == user_id)
    if action is not None:
        query = query.filter(AuditLog.action == action)
    if entity_type is not None:
        query = query.filter(AuditLog.entity_type == entity_type)
    if date_from is not None:
        query = query.filter(AuditLog.created_at >= date_from)
    if date_to is not None:
        query = query.filter(AuditLog.created_at <= date_to)

    total = query.count()
    total_pages = math.ceil(total / page_size) if total > 0 else 1

    logs = (
        query.order_by(AuditLog.created_at.desc())
        .offset((page - 1) * page_size)
        .limit(page_size)
        .all()
    )

    # Enrich with username
    user_ids = {log.user_id for log in logs if log.user_id}
    users_map = {}
    if user_ids:
        users = db.query(User).filter(User.id.in_(user_ids)).all()
        users_map = {u.id: u.username for u in users}

    items = []
    for log in logs:
        item = AuditLogResponse(
            id=log.id,
            user_id=log.user_id,
            username=users_map.get(log.user_id),
            action=log.action,
            entity_type=log.entity_type,
            entity_id=log.entity_id,
            details=log.details,
            ip_address=log.ip_address,
            created_at=log.created_at,
        )
        items.append(item)

    return AuditLogListResponse(
        items=items,
        total=total,
        page=page,
        page_size=page_size,
        total_pages=total_pages,
    )

File: frontend/src/api/auditLogs.ts
import { useQuery } from "@tanstack/react-query";
import apiClient from "./client";
import type { AuditLog } from "../types/user";
import type { PaginatedResponse } from "../types/common";

interface AuditLogParams {
  page?: number;
  page_size?: number;
  user_id?: number;
  action?: string;
  entity_type?: string;
  date_from?: string;
  date_to?: string;
}

const KEYS = {
  all: ["audit-logs"] as const,
  lists: () => [...KEYS.all, "list"] as const,
  list: (params: AuditLogParams) => [...KEYS.lists(), params] as const,
};

async function getAuditLogs(params: AuditLogParams = {}) {
  const { data } = await apiClient.get<PaginatedResponse<AuditLog>>(
    "/audit-logs",
    { params }
  );
  return data;
}

export function useAuditLogs(params: AuditLogParams = {}) {
  return useQuery({
    queryKey: KEYS.list(params),
    queryFn: () => getAuditLogs(params),
  });
}
